### PR TITLE
Add restrictions of object retention to AWS S3 limits

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add restrictions of object retention to AWS S3 limits

## Motivation and Context
This PR also fixes issues related

 - Peer notification handler was missing "/"
 - Missing Prometheus metrics for retention APIs

## How to test this PR?
Using `mc lock` command

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#b8404658-ad19-4dd8-af80-b27cb2055eec) |
| mint-compression-fs.sh | :x: [more...](#dabefe35-3228-4cb0-96ee-f4e956a4d67f) |
<a name="b8404658-ad19-4dd8-af80-b27cb2055eec"></a>
### 8514-7dd91d5/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:31215
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4567033d0655:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 3 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 49 seconds
(4/14) Running aws-sdk-ruby tests ... done in 16 seconds
(5/14) Running awscli tests ... done in 2 minutes and 16 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 12 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 19 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 50 seconds
{
  "args": {},
  "duration": 87091,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
<a name="dabefe35-3228-4cb0-96ee-f4e956a4d67f"></a>
### 8514-7dd91d5/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:30784
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c43588a087de:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 1 minutes and 56 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 44 seconds
(8/14) Running minio-dotnet tests ... done in 25 seconds
(9/14) Running minio-go tests ... done in 49 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 24 seconds
(11/14) Running minio-js tests ... done in 43 seconds
(12/14) Running minio-py tests ... FAILED in 2 minutes and 26 seconds
{
  "args": {
    "response._body": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Request has expired</Message><Key>c415b40d-e462-4eda-b78a-847b43ffaf50</Key><BucketName>minio-py-test-c7df64cc-e190-47d3-bf19-d79f4222ca46</BucketName><Resource>/minio-py-test-c7df64cc-e190-47d3-bf19-d79f4222ca46/c415b40d-e462-4eda-b78a-847b43ffaf50</Resource><RequestId>15D690702403C8D4</RequestId><HostId>75e14585-f2fa-4cdd-8e65-6936abf71f48</HostId></Error>",
    "response.status": 403,
    "bucket_name": "minio-py-test-c7df64cc-e190-47d3-bf19-d79f4222ca46",
    "response.reason": "Forbidden",
    "response.headers": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"440\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/xml\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-11-12T23-43-38Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15D690702403C8D4\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Wed, 13 Nov 2019 00:12:11 GMT\"]}}",
    "object_name": "c415b40d-e462-4eda-b78a-847b43ffaf50",
    "presigned_get_object_url": "http://72.28.97.51:30784/minio-py-test-c7df64cc-e190-47d3-bf19-d79f4222ca46/c415b40d-e462-4eda-b78a-847b43ffaf50?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20191113%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191113T001155Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&X-Amz-Signature=4ae9146b9aa4e269f5f6ee518cf72a4cd15ccdba75bcb7ecad9e8f1f5cf43ae9"
  },
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "message": "AccessDenied: message: Access Denied",
  "status": "FAIL",
  "duration": 63,
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1345, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2040, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1359, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "name": "minio-py:test_presigned_get_object_expiry_5sec"
}

Executed 11 out of 14 tests successfully.
```
Error: No such image: minio/minio:8514-7dd91d5

**Review by kannappanr — DISMISSED** [review]
LGTM

**Review by kannappanr — DISMISSED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#5af6f772-99f3-48d5-aef2-dc1bc523b205) |
<a name="5af6f772-99f3-48d5-aef2-dc1bc523b205"></a>
### 8514-7032d7f/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:31466
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c576d960c7f7:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 3 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 54 seconds
(4/14) Running aws-sdk-ruby tests ... done in 17 seconds
(5/14) Running awscli tests ... done in 2 minutes and 38 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 58 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 32 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 33 seconds
{
  "args": {},
  "duration": 44297,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Error: No such image: minio/minio:8514-7032d7f

File: cmd/object-lock.go [hunk 1]
```diff
@@ -120,10 +127,19 @@ func (dr *DefaultRetention) UnmarshalXML(d *xml.Decoder, start xml.StartElement)
...
 	if retention.Days != nil {
 		if *retention.Days == 0 {
-			return fmt.Errorf("Days should not be zero")
+			return fmt.Errorf("Default retention period must be a positive integer value for 'Days'")
+		}
+		if *retention.Days > maximumRetentionDays {
+			return fmt.Errorf("Default retention period too large for 'Days' %w", *retention.Days)
+		}
+	}
+	if retention.Years != nil {
```
Comment (balamurugana) [inline cmd/object-lock.go]: you can cut down this `if` by `else`

Comment (harshavardhana) [reply cmd/object-lock.go]: done

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#55bda550-7946-47da-a5e1-d63bba9f3fb7) |
| mint-compression-fs.sh | :x: [more...](#216fab10-f941-4edc-8f23-a523c6419641) |
<a name="55bda550-7946-47da-a5e1-d63bba9f3fb7"></a>
### 8514-d430b6a/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:32537
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 079810a8b32a:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 56 seconds
(4/14) Running aws-sdk-ruby tests ... done in 15 seconds
(5/14) Running awscli tests ... done in 2 minutes and 24 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 2 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 30 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 51 seconds
{
  "args": {},
  "duration": 94076,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
<a name="216fab10-f941-4edc-8f23-a523c6419641"></a>
### 8514-d430b6a/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:31626
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d8267bee46ac:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 6 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 1 seconds
(5/14) Running awscli tests ... done in 1 minutes and 0 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 22 seconds
(8/14) Running minio-dotnet tests ... done in 23 seconds
(9/14) Running minio-go tests ... done in 57 seconds
(10/14) Running minio-java tests ... done in 18 seconds
(11/14) Running minio-js tests ... FAILED in 6 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 11,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:83:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:153:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:77:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:607:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:616:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:629:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:567:22) at IncomingMessage.onend (_stream_readable.js:514:10) at endReadableNT (_stream_readable.js:978:12) at _combinedTickCallback (internal/process/next_tick.js:80:11) at process._tickCallback (internal/process/next_tick.js:104:9)"
}

Executed 10 out of 14 tests successfully.
```
Error: No such image: minio/minio:8514-d430b6a

**Review by klauspost — APPROVED** [review]
LGTM

**Review by kannappanr — APPROVED** [review]

**Pull request merged**